dbms/src/Common/UTF8Helpers.h

```diff
@@ -49,9 +52,37 @@ inline size_t seqLength(const UInt8 first_octet)
 inline size_t countCodePoints(const UInt8 * data, size_t size)
 {
     size_t res = 0;
+    const auto end = data + size;
 
-    /// TODO SIMD implementation looks quite simple.
-    for (auto end = data + size; data < end; ++data) /// Skip UTF-8 continuation bytes.
+#if __SSE2__
+    const auto bytes_sse = sizeof(__m128i);
+    const auto src_end_sse = (data + size) - (size % bytes_sse);
+
+    const auto upper_bound = _mm_set1_epi8(0x7F + 1);
+    const auto lower_bound = _mm_set1_epi8(0xC0 - 1);
+
+    for (; data < src_end_sse;)
+    {
+        UInt8 mem_res[16] = {0};
+        auto sse_res = _mm_set1_epi8(0);
+
+        for (int i = 0; i < 0XFF && data < src_end_sse; ++i, data += bytes_sse)
+        {
+            const auto chars = _mm_loadu_si128(reinterpret_cast<const __m128i *>(data));
+            sse_res = _mm_add_epi8(sse_res,
+                                   _mm_or_si128(_mm_cmplt_epi8(chars, upper_bound),
```

Comment:
It's interesting if we can do subtract (add) with overflow (to align a range with zero) and then do just single comparison.